### PR TITLE
Don't block for all bytes in network receive

### PR DESCRIPTION
**Pull request by gordonwang0**
Description
-----------

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [x] My code is Linted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: libraries/abstractions/platform/freertos/iot_network_freertos.c [hunk 1]
```diff
@@ -524,32 +524,18 @@ size_t IotNetworkAfr_Receive( void * pConnection,
...
-        {
-            IotLogError( "Error %ld while receiving data.", ( long int ) socketStatus );
-            break;
-        }
-        else
-        {
-            bytesReceived += ( size_t ) socketStatus;
-            bytesRemaining -= ( size_t ) socketStatus;
+    socketStatus = SOCKETS_Recv( pNetworkConnection->socket,
+                                 pBuffer + bytesReceived,
```
Comment (aggarg) [inline libraries/abstractions/platform/freertos/iot_network_freertos.c]: After this simplification, we can get rid of ```bytesRemaining ```. 
```
socketStatus = SOCKETS_Recv( pNetworkConnection->socket,
                                     pBuffer,
                                     bytesRequested,
                                     0 );
```
```
    if( socketStatus <= 0 )
    {
        IotLogError( "Error %ld while receiving data.", ( long int ) socketStatus );
    }
    else
    {
        bytesReceived = ( size_t ) socketStatus;
    }
```

File: libraries/c_sdk/standard/mqtt/src/iot_mqtt_network.c [hunk 2]
```diff
@@ -232,11 +232,23 @@ static IotMqttError_t _getIncomingPacket( void * pNetworkConnection,
...
+                break;
+            }
+            else
+            {
+                totalBytesRead += bytesRead;
+            }
+        }
 
-        if( dataBytesRead != pIncomingPacket->remainingLength )
+        if( totalBytesRead != pIncomingPacket->remainingLength )
```
Comment (aggarg) [inline libraries/c_sdk/standard/mqtt/src/iot_mqtt_network.c]: May be add a log message saying that not complete message was received.

File: libraries/c_sdk/standard/mqtt/src/iot_mqtt_network.c [hunk 3]
```diff
@@ -232,11 +232,23 @@ static IotMqttError_t _getIncomingPacket( void * pNetworkConnection,
...
-        dataBytesRead = pMqttConnection->pNetworkInterface->receive( pNetworkConnection,
-                                                                     pIncomingPacket->pRemainingData,
-                                                                     pIncomingPacket->remainingLength );
+        while( totalBytesRead < pIncomingPacket->remainingLength )
+        {
+            bytesRead = pMqttConnection->pNetworkInterface->receive( pNetworkConnection,
+                                                                     pIncomingPacket->pRemainingData + totalBytesRead,
+                                                                     pIncomingPacket->remainingLength - totalBytesRead );
+
+            if( bytesRead == 0 )
```
Comment (ravibhagavandas) [inline libraries/c_sdk/standard/mqtt/src/iot_mqtt_network.c]: For non blocking sockets, `bytesRead` could be 0, if there is no data currently on socket ( EWOULDBLOCK ).  Previous code handles this within the while loop of `IotNetworkAfr_Receive()`

**Comment by cobusve** [comment]
Please take care before you merge this PR. Changing the semantics of the receive function like this will at least break the MQTT code.

**Comment by cobusve** [comment]
From what I can see the users of this receive function depend on the function blocking as described in the header. Besides, if we change the function to not block then we should at the very least also update the header to no longer make this claim, and of course update all the parts of the code which was written depending on this to handler partial reads reliably.

File: libraries/c_sdk/standard/mqtt/src/iot_mqtt_network.c (hunk 3 above)
Comment (sarenameas) [reply libraries/c_sdk/standard/mqtt/src/iot_mqtt_network.c]: I believe a non-blocking socket returns EWOULDBLOCK while the blocking sockets return 0 on timeout.

Perhaps we need to return socket layers after all, so the application can ask again? 
Suggestion:
```c
/* param[in,out] bytesRequested */
IotNetworkError_t ( * receive )( void * pConnection,
                           uint8_t * pBuffer,
                           size_t* bytesRequested );
```

In the old code, when 0 was returned from SOCKETS_Recv() the function IotNetwokrAfr() did not loop again. It broke out and returned with the possible partial data.

Comment (cobusve) [reply libraries/c_sdk/standard/mqtt/src/iot_mqtt_network.c]: My main objection is that the design matches the documentation which states specifically the following:

/**
     * @brief Block and wait for incoming network data.
     *
     * Wait for a message of size `bytesRequested` to arrive on the network and
     * place it in `pBuffer`.
     *
     * @param[in] pConnection The connection to wait on, defined by the network
     * stack.
     * @param[out] pBuffer Where to place the incoming network data. This buffer
     * must be at least `bytesRequested` in size.
     * @param[in] bytesRequested How many bytes to wait for. `pBuffer` must be at
     * least this size.
     *
     * @return The number of bytes successfully received. This should be
     * `bytesRequested` when successful. Any other value may indicate an error.
     */

People (as they should) rely on the fact that the port implements what it promises. Looking at the MQTT code e.g. here seems to prove that this is at least true in MQTT.


       dataBytesRead = pMqttConnection->pNetworkInterface->receive( pNetworkConnection,
                                                                     pIncomingPacket->pRemainingData,
                                                                     pIncomingPacket->remainingLength );

        if( dataBytesRead != pIncomingPacket->remainingLength )
        {
            IOT_SET_AND_GOTO_CLEANUP( IOT_MQTT_BAD_RESPONSE );
        }
        else
        {
            EMPTY_ELSE_MARKER;
        }

Comment (cobusve) [reply libraries/c_sdk/standard/mqtt/src/iot_mqtt_network.c]: This means that you would always have a race condition in MQTT and if you break the packet in half by failing out it looks like bad things will happen after this ...

**Comment by sarenameas** [comment]
> Please take a look at everywhere this function is being used e.g. in MQTT in file libraries/c_sdk/standard/mqtt/src/iot_mqtt_network.c at line 235. This code will fail if you read part of a packet, which will happen when networks have mismatched MTU's or packets get fragmented on IP layers.
> 

Please correct me if I am wrong, but from what I see, this possibility still existed with the old implementation.
On the non-CC3220SF blocking sockets (other boards basically), we set a receive timeout once in IotNetworkAfr_Create() to IOT_NETWORK_SOCKET_POLL_MS which is defaulted to 1000 ms. 
If the socket timed-out after only delivering a part of the packet, some of the ported SOCKETS_Recv would return a 0. If a zero was returned, then the older implementation returned right away. So the user needed to do something about the partial packet possibility anyways. 

```c
        else if( socketStatus <= 0 )
        {
            IotLogError( "Error %ld while receiving data.", ( long int ) socketStatus );
            break;
        }
```

Only in the case of EWOULDBLOCK would the loop continue. 

The functionality of the network receive in MQTT is not changed from previous behavior with this update, except for the lack for the loop continuing on a EWOULDBLOCK returned. 

I am not saying that since there is not really any change from the old that things are correct. We should definitely be wary of partial packets and have way to handle them.

**Pull request closed without merging**